src/predictor.rs

```diff
@@ -0,0 +1,784 @@
+            for i in (samples * 8..buf.len()).step_by(8) {
+                let v = u64::from_ne_bytes(buf[i..][..8].try_into().unwrap());
+                let p = u64::from_ne_bytes(buf[i - 8 * samples..][..8].try_into().unwrap());
+                buf[i..][..8].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        _ => {
+            unreachable!("Caller should have validated arguments. Please file a bug.")
+        }
+    }
+}
+
+/// Fix endianness. If `byte_order` matches the host, then conversion is a no-op.
+///
+/// from image-tiff
+pub fn fix_endianness(buffer: Bytes, byte_order: Endianness, bit_depth: u16) -> Bytes {
+    match byte_order {
+        #[cfg(target_endian = "little")]
+        Endianness::LittleEndian => buffer,
+        #[cfg(target_endian = "big")]
+        Endianness::LittleEndian => {
+            let mut buf = BytesMut::from(buffer);
+            match bit_depth {
+                0..=8 => {}
+                9..=16 => buf.chunks_exact_mut(2).for_each(|v| {
+                    v.copy_from_slice(&u16::from_le_bytes((*v).try_into().unwrap()).to_ne_bytes())
+                }),
+                17..=32 => buf.chunks_exact_mut(4).for_each(|v| {
+                    v.copy_from_slice(&u32::from_le_bytes((*v).try_into().unwrap()).to_ne_bytes())
+                }),
+                _ => buf.chunks_exact_mut(8).for_each(|v| {
+                    v.copy_from_slice(&u64::from_le_bytes((*v).try_into().unwrap()).to_ne_bytes())
+                }),
+            }
+            buf.freeze()
+        }
+        #[cfg(target_endian = "little")]
+        Endianness::BigEndian => {
+            let mut buf = BytesMut::from(buffer);
+            match bit_depth {
+                0..=8 => {}
+                9..=16 => buf.chunks_exact_mut(2).for_each(|v| {
+                    v.copy_from_slice(&u16::from_be_bytes((*v).try_into().unwrap()).to_ne_bytes())
+                }),
+                17..=32 => buf.chunks_exact_mut(4).for_each(|v| {
+                    v.copy_from_slice(&u32::from_be_bytes((*v).try_into().unwrap()).to_ne_bytes())
+                }),
+                _ => buf.chunks_exact_mut(8).for_each(|v| {
+                    v.copy_from_slice(&u64::from_be_bytes((*v).try_into().unwrap()).to_ne_bytes())
+                }),
+            };
+            buf.freeze()
+        }
+        #[cfg(target_endian = "big")]
+        Endianness::BigEndian => buffer,
+    }
+}
+
+/// Floating point predictor
+#[derive(Debug)]
+pub struct FloatingPointPredictor;
+
+impl Unpredict for FloatingPointPredictor {
+    fn fix_endianness_and_unpredict(
+        &self,
+        buffer: Bytes,
+        predictor_info: &PredictorInfo,
+        tile_x: u32,
+        tile_y: u32,
+    ) -> AsyncTiffResult<Bytes> {
+        let output_row_stride = predictor_info.output_row_stride(tile_x)?;
+        let mut res: BytesMut =
+            BytesMut::zeroed(output_row_stride * predictor_info.output_rows(tile_y)?);
```

Comment:
Here now I did an output_rows, which is more if we have `Planar` planar config. I thought then at least we can give the user the data and they can split it themselves?

Reply:
So now the `Planar` output will be:
```
[
Red...,
Green.,
Blue..,
]
```
where each band is `chunk_width_pixels()*chunk_height_pixels()`

Reply:
I thought this made sense, because the decoder (decompression) step decodes the entire data, which is like structured like this